### ISSUE TITLE
FileAndResourceDirectives.getFromResourceDirectory with sub directories may list files 

### ISSUE THREAD

**Issue by matlockx** [akka/akka#17495]
When passing a directory which contains subdirectories to FileAndResourceDirectives.getFromResource, request which map to a subdirectory will create a response with the file listing of that subdirectory. 
In detail: When the path of the request ends with a '/' the request will be rejected, if not the request comes through and the file listing will be returned.

**Comment by jrudolph** [comment]
Can you post relevant parts of your route? Do you really mean `getFromResource` or `getFromResourceDirectory`?

**Comment by jrudolph** [comment]
I can remember that we discussed this issue in the context of spray a long time ago. The problem is that it's possible to use `ClassLoader.getResource(dirName).openStream` to list contents of classpath resource directories (if the classpath entry is a file system directory). It doesn't seem to work when the classpath entry is a jar.

I guess the fix is then to check if the URL returned from `ClassLoader.getResource` is a `file` URL, and if yes, check that it doesn't point to a directory.

/cc @sirthias

**Comment by matlockx** [comment]
Actually you are right, it's getFromResourceDirectory. Sorry for that.

The path is something like:

```scala
    pathPrefix("game") {
	    	get {
                  getFromResourceDirectory("public")
                }
   }
```

**Comment by jrudolph** [comment]
Just to confirm: are you experiencing this when everything is packaged into a jar or does this happen during development (e.g. with sbt) when the referenced resources are in a classpath directory?

**Comment by matlockx** [comment]
I get it during development with sbt. 
I just checked it in production and there it's a NPE.... 

ava.lang.NullPointerException: null
        at java.io.FilterInputStream.close(FilterInputStream.java:181) ~[na:1.8.0_40]
        at sun.net.www.protocol.jar.JarURLConnection$JarURLInputStream.close(JarURLConnection.java:108) ~[na:1.8.0_40]
        at spray.routing.directives.FileAndResourceDirectives$class.getFromResource(FileAndResourceDirectives.scala:128) ~[io.spray.spray-routing_2.11-1.3.2.jar:na]
        at iwg.api.routes.StaticGameContentRoute.getFromResource(StaticGameContentRoutes.scala:15) ~[io.lotteries.iwg-ticketpool_2.11-0.4.17.jar:0.4.17]
        at spray.routing.directives.FileAndResourceDirectives$class.getFromResource(FileAndResourceDirectives.scala:104) ~[io.spray.spray-routing_2.11-1.3.2.jar:na]

**Comment by jrudolph** [comment]
FTR The PR that fixed this was #17503.
